postgresql: implement opt-in JIT support

Ma27 opened the pull request:
###### Description of changes

Closes #150801

Note: I decided against resuming directly on #150801 because the conflict was too big (and resolving it seemed too error-prone to me). Also the `this`-refactoring could be done in an easier manner, i.e. by exposing JIT attributes with the correct configuration. More on that below.

This patch creates variants of the `postgresql*`-packages with JIT[1] support. Please note that a lot of the work was derived from previous patches filed by other contributors, namely dasJ, andir and abbradar, hence the co-authored-by tags below.

Effectively, the following things have changed:

* For JIT variants an LLVM-backed stdenv with clang is now used as
  suggested by dasJ[2]. We need LLVM and CLang[3] anyways to build the
  JIT-part, so no need to mix this up with GCC's stdenv. Also, using the
  `dev`-output of LLVM and clang's stdenv for building (and adding llvm
  libs as build-inputs) seems more cross friendly to me (which will
  become useful when cross-building for JIT-variants will actually be
  supported).

* Plugins inherit the build flags from the Makefiles in `$out/lib/pgxs/src` (e.g. `-Werror=unguarded-availability-new`). Since some of the flags are clang-specific (and stem from the use of the CLang stdenv) and don't work on gcc, the stdenv of `pkgs.postgresql` is passed to the plugins. I.e., plugins for non-JIT variants are built with a gcc stdenv on Linux and plugins for JIT variants with a clang stdenv.

  Since `plv8` hard-codes `gcc` as `$CC` in its Makefile[4], I marked it as broken for JIT-variants of postgresql only.

* Added a test-matrix to confirm that JIT works fine on each `pkgs.postgresql_*_jit` (thanks Andi for the original test in #124804!).

* For each postgresql version, a new attribute `postgresql_<version>_jit` (and a corresponding `postgresqlPackages<version>JitPackages`) are now exposed for better discoverability and prebuilt artifacts in the binary cache.

* In #150801 the `this`-argument was replaced by an internal recursion. I decided against this approach because it'd blow up the diff even more which makes the readability way harder and also harder to revert this if necessary.

  Instead, it is made sure that `this` always points to the correct variant of `postgresql` and re-using that in an additional `.override {}`-expression is trivial because the JIT-variant is exposed in `all-packages.nix`.

* I think the changes are sufficiently big to actually add myself as maintainer here.

* Added `libxcrypt` to `buildInputs` for versions <v13. While building things with an LLVM stdenv, these versions complained that the extern `crypt()` symbol can't be found. Not sure what this is exactly about, but since we want to switch to libxcrypt for `crypt()` usage anyways[5] I decided to add it. For >=13 it's not relevant anymore anyways[6].

* JIT support doesn't work with cross-compilation. It is attempted to build LLVM-bytecode
  (`%.bc` is the corresponding `make(1)`-rule) for each sub-directory in `backend/` for
  the JIT apparently, but with a $(CLANG) that can produce binaries for the build, not the
  host-platform.

  I managed to get a cross-build with JIT support working with
  `depsBuildBuild = [ llvmPackages.clang ] ++ buildInputs`, but considering that the
  resulting LLVM IR isn't platform-independent this doesn't give you much.
  In fact, I tried to test the result in a VM-test, but as soon as JIT was used to optimize
  a query, postgres would coredump with `Illegal instruction`.

A common concern of the original approach - with llvm as build input - was the massive increase of closure size. With the new approach of using the LLVM stdenv directly and patching out references to the clang drv in `$out` the effective closure size changes are:

    $ nix path-info -Sh $(nix-build -A postgresql_14)
    /nix/store/kssxxqycwa3c7kmwmykwxqvspxxa6r1w-postgresql-14.7	306.4M
    $ nix path-info -Sh $(nix-build -A postgresql_14_jit)
    /nix/store/xc7qmgqrn4h5yr4vmdwy56gs4bmja9ym-postgresql-14.7	689.2M

Most of the increase in closure-size stems from the `lib`-output of LLVM

    $ nix path-info -Sh /nix/store/5r97sbs5j6mw7qnbg8nhnq1gad9973ap-llvm-11.1.0-lib
    /nix/store/5r97sbs5j6mw7qnbg8nhnq1gad9973ap-llvm-11.1.0-lib	349.8M

which is why this shouldn't be enabled by default.

While this is quite much because of LLVM, it's still a massive improvement over the simple approach of adding llvm/clang as build-inputs and building with `--with-llvm`:

    $ nix path-info -Sh $(nix-build -E '
	with import ./. {};
	postgresql.overrideAttrs ({ configureFlags ? [], buildInputs ? [], ... }: {
	  configureFlags = configureFlags ++ [ "--with-llvm" ];
	  buildInputs = buildInputs ++ [ llvm clang ];
	})' -j0)
    /nix/store/i3bd2r21c6c3428xb4gavjnplfqxn27p-postgresql-14.7	  1.6G

Co-authored-by: Andreas Rammhold <andreas@rammhold.de>
Co-authored-by: Janne Heß <janne@hess.ooo>
Co-authored-by: Nikolay Amiantov <ab@fmap.me>

[1] https://www.postgresql.org/docs/current/jit-reason.html [2] https://github.com/NixOS/nixpkgs/pull/124804#issuecomment-864616931
    & https://github.com/NixOS/nixpkgs/pull/150801#issuecomment-1467868321
[3] This fails with the following error otherwise:
    ```
    configure: error: clang not found, but required when compiling --with-llvm, specify with CLANG=
    ```
[4] https://github.com/plv8/plv8/blob/v3.1.5/Makefile#L14 [5] https://github.com/NixOS/nixpkgs/pull/181764
[6] https://github.com/postgres/postgres/commit/c45643d618e35ec2fe91438df15abd4f3c0d85ca

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Ma27 committed:
postgresql: implement opt-in JIT support

Closes #150801

Note: I decided against resuming directly on #150801 because the
conflict was too big (and resolving it seemed too error-prone to me).
Also the `this`-refactoring could be done in an easier manner, i.e. by
exposing JIT attributes with the correct configuration. More on that
below.

This patch creates variants of the `postgresql*`-packages with JIT[1]
support. Please note that a lot of the work was derived from previous
patches filed by other contributors, namely dasJ, andir and abbradar,
hence the co-authored-by tags below.

Effectively, the following things have changed:

* For JIT variants an LLVM-backed stdenv with clang is now used as
  suggested by dasJ[2]. We need LLVM and CLang[3] anyways to build the
  JIT-part, so no need to mix this up with GCC's stdenv. Also, using the
  `dev`-output of LLVM and clang's stdenv for building (and adding llvm
  libs as build-inputs) seems more cross friendly to me (which will
  become useful when cross-building for JIT-variants will actually be
  supported).

* Plugins inherit the build flags from the Makefiles in
  `$out/lib/pgxs/src` (e.g. `-Werror=unguarded-availability-new`). Since
  some of the flags are clang-specific (and stem from the use of the
  CLang stdenv) and don't work on gcc, the stdenv of `pkgs.postgresql`
  is passed to the plugins. I.e., plugins for non-JIT variants are built
  with a gcc stdenv on Linux and plugins for JIT variants with a clang
  stdenv.

  Since `plv8` hard-codes `gcc` as `$CC` in its Makefile[4], I marked it
  as broken for JIT-variants of postgresql only.

* Added a test-matrix to confirm that JIT works fine on each
  `pkgs.postgresql_*_jit` (thanks Andi for the original test in
  #124804!).

* For each postgresql version, a new attribute
  `postgresql_<version>_jit` (and a corresponding
  `postgresqlPackages<version>JitPackages`) are now exposed for better
  discoverability and prebuilt artifacts in the binary cache.

* In #150801 the `this`-argument was replaced by an internal recursion.
  I decided against this approach because it'd blow up the diff even
  more which makes the readability way harder and also harder to revert
  this if necessary.

  Instead, it is made sure that `this` always points to the correct
  variant of `postgresql` and re-using that in an additional
  `.override {}`-expression is trivial because the JIT-variant is
  exposed in `all-packages.nix`.

* I think the changes are sufficiently big to actually add myself as
  maintainer here.

* Added `libxcrypt` to `buildInputs` for versions <v13. While
  building things with an LLVM stdenv, these versions complained that
  the extern `crypt()` symbol can't be found. Not sure what this is
  exactly about, but since we want to switch to libxcrypt for `crypt()`
  usage anyways[5] I decided to add it. For >=13 it's not relevant
  anymore anyways[6].

* JIT support doesn't work with cross-compilation. It is attempted to
  build LLVM-bytecode (`%.bc` is the corresponding `make(1)`-rule) for
  each sub-directory in `backend/` for the JIT apparently, but with a
  $(CLANG) that can produce binaries for the build, not the host-platform.

  I managed to get a cross-build with JIT support working with
  `depsBuildBuild = [ llvmPackages.clang ] ++ buildInputs`, but
  considering that the resulting LLVM IR isn't platform-independent this
  doesn't give you much. In fact, I tried to test the result in a VM-test,
  but as soon as JIT was used to optimize a query, postgres would
  coredump with `Illegal instruction`.

A common concern of the original approach - with llvm as build input -
was the massive increase of closure size. With the new approach of using
the LLVM stdenv directly and patching out references to the clang drv in
`$out` the effective closure size changes are:

    $ nix path-info -Sh $(nix-build -A postgresql_14)
    /nix/store/kssxxqycwa3c7kmwmykwxqvspxxa6r1w-postgresql-14.7	306.4M
    $ nix path-info -Sh $(nix-build -A postgresql_14_jit)
    /nix/store/xc7qmgqrn4h5yr4vmdwy56gs4bmja9ym-postgresql-14.7	689.2M

Most of the increase in closure-size stems from the `lib`-output of
LLVM

    $ nix path-info -Sh /nix/store/5r97sbs5j6mw7qnbg8nhnq1gad9973ap-llvm-11.1.0-lib
    /nix/store/5r97sbs5j6mw7qnbg8nhnq1gad9973ap-llvm-11.1.0-lib	349.8M

which is why this shouldn't be enabled by default.

While this is quite much because of LLVM, it's still a massive
improvement over the simple approach of adding llvm/clang as
build-inputs and building with `--with-llvm`:

    $ nix path-info -Sh $(nix-build -E '
	with import ./. {};
	postgresql.overrideAttrs ({ configureFlags ? [], buildInputs ? [], ... }: {
	  configureFlags = configureFlags ++ [ "--with-llvm" ];
	  buildInputs = buildInputs ++ [ llvm clang ];
	})' -j0)
    /nix/store/i3bd2r21c6c3428xb4gavjnplfqxn27p-postgresql-14.7	  1.6G

Co-authored-by: Andreas Rammhold <andreas@rammhold.de>
Co-authored-by: Janne Heß <janne@hess.ooo>
Co-authored-by: Nikolay Amiantov <ab@fmap.me>

[1] https://www.postgresql.org/docs/current/jit-reason.html
[2] https://github.com/NixOS/nixpkgs/pull/124804#issuecomment-864616931
    & https://github.com/NixOS/nixpkgs/pull/150801#issuecomment-1467868321
[3] This fails with the following error otherwise:
    ```
    configure: error: clang not found, but required when compiling --with-llvm, specify with CLANG=
    ```
[4] https://github.com/plv8/plv8/blob/v3.1.5/Makefile#L14
[5] https://github.com/NixOS/nixpkgs/pull/181764
[6] https://github.com/postgres/postgres/commit/c45643d618e35ec2fe91438df15abd4f3c0d85ca

Ma27 commented:
@ofborg build postgresql_jit

Ma27 commented:
So, darwin is failing for the JIT variant, but appears to be working fine for the non-JIT version, so existing users won't be affected.

I guess the issue is fixable rather trivially, but I don't own a mac and I don't want to pay cloud-providers for that (this is something I implemented in my spare-time!), so I'm inclined to mark it as broken and wait for someone who's interested and has the resources to fix it unless somebody can give me a darwin environment.

mweinelt commented:
There exists a community-maintained darwin builder at https://github.com/winterqt/darwin-build-box.

RaitoBezarius commented:
Testing this on a PostgreSQL-based Nextcloud personal production.

RaitoBezarius reviewed (commented):
It would be interesting if there was a way to enable JIT in the NixOS module without having to find out the existing PostgreSQL version wrt to stateVersion.

RaitoBezarius reviewed (approved):
It didn't break NextCloud yet. I will try to measure speed improvements wrt to queries, including for stuff like Mastodon.

Ma27 commented:
> It would be interesting if there was a way to enable JIT in the NixOS module without having to find out the existing PostgreSQL version wrt to stateVersion

it should be relatively trivial considering that we generate `_jit`-derivations out of the existing postgrtesql packages: https://github.com/NixOS/nixpkgs/pull/221851/files#diff-b2d51596507a2d6c94ff9a31c0b3fab6b254d2589bb802d92fbcc1221e876b10R331-R340

Just didn't think it'd be actually needed, but I guess that's just me pinning my postgres manually :) 
Will update tomorrow.

Ma27 committed:
postgresql: pass through JIT-enabled variant of non-JIT postgres and vice versa

This is useful if your postgresql version is dependant on
`system.stateVersion` and not pinned down manually. Then it's not
necessary to find out which version exactly is in use and define
`package` manually, but just stay with what NixOS provides as default:

    $ nix-instantiate -A postgresql
    /nix/store/82fzmb77mz2b787dgj7mn4a8i4f6l6sn-postgresql-14.7.drv
    $ nix-instantiate -A postgresql_jit
    /nix/store/qsjkb72fcrrfpsszrwbsi9q9wgp39m50-postgresql-14.7.drv
    $ nix-instantiate -A postgresql.withJIT
    /nix/store/qsjkb72fcrrfpsszrwbsi9q9wgp39m50-postgresql-14.7.drv
    $ nix-instantiate -A postgresql.withJIT.withoutJIT
    /nix/store/82fzmb77mz2b787dgj7mn4a8i4f6l6sn-postgresql-14.7.drv

I.e. you can use postgresql with JIT (for complex queries only[1]) like
this:

    services.postgresql = {
      enable = true;
      enableJIT = true;
    };

Performing a new override instead of re-using the `_jit`-variants for
that has the nice property that overlays for the original package apply
to the JIT-enabled variant, i.e.

    with import ./. {
      overlays = [
        (self: super: {
          postgresql = super.postgresql.overrideAttrs (_: { fnord = "snens"; });
        })
      ];
    };
    postgresql.withJIT.fnord

still gives the string `snens` whereas `postgresql_jit` doesn't have the
attribute `fnord` in its derivation.

[1] https://www.postgresql.org/docs/current/runtime-config-query.html#GUC-JIT-ABOVE-COST

Ma27 committed:
nixos/tests/postgresql: fix deprecation warning

RaitoBezarius commented:
Awesome, I really like it.

RaitoBezarius commented:
I would pester you just to add this little change in the release notes saying that postgresql has a `withJIT` passthru useful when you have pgsql package but you don't know the version exactly.

Ma27 commented:
@RaitoBezarius done!

RaitoBezarius commented:
If I'm not wrong, all is left in this PR:

- [ ] Fix Darwin.
- [ ] Retarget staging (I believe)?
- [ ] Last thorough review

Then we go for the merge?

Ma27 committed:
postgresql_jit: fix darwin build

So, patchelf works on ELF files and these don't exist on Darwin. I'm not
aware of any other way to eliminate all references of the executable's
body while preserving library paths in the header. This still works, the
only problem is that darwin has a dependency on llvm.dev, so it's
runtime closure is larger.

Ma27 commented:
@RaitoBezarius I guess reviews will be needed by other folks ;-)
(cc @marsam )

Fixed the darwin build in a rather dirty manner: https://github.com/NixOS/nixpkgs/pull/221851/commits/36a1fae48a87d81141f1a1e59be0a92f8b35b20d (see commit message for details).

Ma27 commented:
also pinging @dasJ @ajs124 for a review who seem to be using this already :)

RaitoBezarius commented:
> @RaitoBezarius I guess reviews will be needed by other folks ;-) (cc @marsam )
> 
> Fixed the darwin build in a rather dirty manner: [36a1fae](https://github.com/NixOS/nixpkgs/commit/36a1fae48a87d81141f1a1e59be0a92f8b35b20d) (see commit message for details).

Certainly, @K900 is giving it a try (or might?).

I redeployed the latest revision in production, no issue so far.

K900 commented:
It's a lot of rebuilds so I probably won't run this consistently, but I am trying to deploy it.

K900 commented:
Gun, meet foot.

Original config:

```
  services.postgresql = {
    enable = true;
    package = pkgs.postgresql_15;
  };
```

New config:

```
  services.postgresql = {
    enable = true;
    enableJIT = true;
    package = pkgs.postgresql_15;
  };
```

JIT doesn't work. 

Correct config:

```
  services.postgresql = {
    enable = true;
    enableJIT = true;
    package = pkgs.postgresql_15_jit;
  };
```

RaitoBezarius commented:
Apologize, I'm partially responsible for nerdsnipping @Ma27 into it. :dagger:

K900 commented:
But one rebuild after, 

```
postgres=# select pg_jit_available();
 pg_jit_available 
------------------
 t
(1 row)
```

K900 commented:
And Synapse feels hugely faster!

Ma27 commented:
Oof, how did I not think of this case sorry! :facepalm: 

Will fix that later :)

SuperSandro2000 commented on pkgs/servers/sql/postgresql/default.nix:
```diff
@@ -99,6 +114,13 @@ let
     postPatch = ''
       # Hardcode the path to pgxs so pg_config returns the path in $out
       substituteInPlace "src/common/config_info.c" --replace HARDCODED_PGXS_PATH "$out/lib"
+
+      ${lib.optionalString jitSupport ''
+        # Force lookup of jit stuff in $out instead of $lib
+        substituteInPlace src/backend/jit/jit.c --replace pkglib_path \"$out/lib\"
+        substituteInPlace src/backend/jit/llvm/llvmjit.c --replace pkglib_path \"$out/lib\"
+        substituteInPlace src/backend/jit/llvm/llvmjit_inline.cpp --replace pkglib_path \"$out/lib\"
+      ''}
     '';
```
```suggestion
      substituteInPlace "src/common/config_info.c" --replace HARDCODED_PGXS_PATH "$out/lib"
    '' + lib.optionalString jitSupport ''
      # Force lookup of jit stuff in $out instead of $lib
      substituteInPlace src/backend/jit/jit.c --replace pkglib_path \"$out/lib\"
      substituteInPlace src/backend/jit/llvm/llvmjit.c --replace pkglib_path \"$out/lib\"
      substituteInPlace src/backend/jit/llvm/llvmjit_inline.cpp --replace pkglib_path \"$out/lib\"
    '';
```

mweinelt replied:
I don't see the point in suggesting formatting changes like this.

SuperSandro2000 replied:
Nesting optionalString instead of appending them is not that nice and causes extra rebuilds.

mweinelt replied:
Can you suggest how inlining vs concating makes a difference regarding rebuilds?

Ma27 replied:
Sorry, my mistake! I confused this suggestion with the isDarwin-part below (where concatenation doesn't make sense).

In fact I already fixed this and because we don't have any trailing newlines in the drv here (in the previous version there was always one newline at the end of `postPatch` which caused a rebuild of the non-JIT packages on this branch which is why this used to point against staging), we actually saved some rebuilds (that's why this PR points to master again).

SuperSandro2000 commented on pkgs/servers/sql/postgresql/default.nix:
```diff
@@ -109,27 +131,56 @@ let
             fi
           done
         fi
+
+        ${lib.optionalString jitSupport ''
+          # Move the bitcode and libllvmjit.so library out of $lib; otherwise, every client that
+          # depends on libpq.so will also have libLLVM.so in its closure too, bloating it
+          moveToOutput "lib/bitcode" "$out"
+          moveToOutput "lib/llvmjit*" "$out"
+
+          # In the case of JIT support, prevent a retained dependency on clang-wrapper
+          substituteInPlace "$out/lib/pgxs/src/Makefile.global" --replace ${self.llvmPackages.stdenv.cc}/bin/clang clang
+          nuke-refs $out/lib/llvmjit_types.bc $(find $out/lib/bitcode -type f)
+
+          # Stop out depending on the default output of llvm
+          substituteInPlace $out/lib/pgxs/src/Makefile.global \
+            --replace ${self.llvmPackages.llvm.out}/bin "" \
+            --replace '$(LLVM_BINPATH)/' ""
+
+          # Stop out depending on the -dev output of llvm
+          substituteInPlace $out/lib/pgxs/src/Makefile.global \
+            --replace ${self.llvmPackages.llvm.dev}/bin/llvm-config llvm-config \
+            --replace -I${self.llvmPackages.llvm.dev}/include ""
+
+          ${lib.optionalString (!stdenv'.isDarwin) ''
+            # Stop lib depending on the -dev output of llvm
+            rpath=$(patchelf --print-rpath $out/lib/llvmjit.so)
+            nuke-refs -e $out $out/lib/llvmjit.so
+            # Restore the correct rpath
+            patchelf $out/lib/llvmjit.so --set-rpath "$rpath"
+          ''}
+        ''}
       '';
```
```suggestion
      '' + lib.optionalString jitSupport ''
        # Move the bitcode and libllvmjit.so library out of $lib; otherwise, every client that
        # depends on libpq.so will also have libLLVM.so in its closure too, bloating it
        moveToOutput "lib/bitcode" "$out"
        moveToOutput "lib/llvmjit*" "$out"

        # In the case of JIT support, prevent a retained dependency on clang-wrapper
        substituteInPlace "$out/lib/pgxs/src/Makefile.global" --replace ${self.llvmPackages.stdenv.cc}/bin/clang clang
        nuke-refs $out/lib/llvmjit_types.bc $(find $out/lib/bitcode -type f)

        # Stop out depending on the default output of llvm
        substituteInPlace $out/lib/pgxs/src/Makefile.global \
          --replace ${self.llvmPackages.llvm.out}/bin "" \
          --replace '$(LLVM_BINPATH)/' ""

        # Stop out depending on the -dev output of llvm
        substituteInPlace $out/lib/pgxs/src/Makefile.global \
          --replace ${self.llvmPackages.llvm.dev}/bin/llvm-config llvm-config \
          --replace -I${self.llvmPackages.llvm.dev}/include ""

        ${lib.optionalString (!stdenv'.isDarwin) ''
          # Stop lib depending on the -dev output of llvm
          rpath=$(patchelf --print-rpath $out/lib/llvmjit.so)
          nuke-refs -e $out $out/lib/llvmjit.so
          # Restore the correct rpath
          patchelf $out/lib/llvmjit.so --set-rpath "$rpath"
        ''}
  
      '';
```

Ma27 replied:
This one is actually reasonable because it should save us a few rebuilds.

RaitoBezarius commented on pkgs/servers/sql/postgresql/default.nix:
```diff
@@ -14,17 +14,28 @@ let
     zstdEnabled = atLeast "15";
 
-  in stdenv.mkDerivation rec {
+    stdenv' = if jitSupport then llvmPackages.stdenv else stdenv;
```
What is your suggestion? Adding an indirection layer seems also to pose some problems I think.

Ma27 replied:
I think that adding even more complexity here is not feasible and IMHO it's the reviewer's job to spot such a mistake. However, we may want to rename `stdenv'` to make it more obvious. Not sure how though, I'm open for suggestions :)

SuperSandro2000 replied:
I don't have one, I just wanted to bring up that the next person is likely first using stdenv

Ma27 replied:
The thing is, I actually tried and I couldn't come up with one, hence the question :)

Ma27 committed:
nixos/postgresql: fix enableJIT

Make sure that JIT is actually available when using

    services.postgresql = {
      enable = true;
      enableJIT = true;
      package = pkgs.postgresql_15;
    };

The current behavior is counter-intuitive because the docs state that
`enableJIT = true;` is sufficient even though it wasn't in that case
because the declared package doesn't have the LLVM dependency.

Fixed by using `package.withJIT` if `enableJIT = true;` and
`package.jitSupport` is `false`.

Also updated the postgresql-jit test to test for that case.

Ma27 committed:
postgresql: save rebuilds of existing packages

...by using `+ lib.optionalString ...` rather than a substitution. That
way the phases don't have additional trailing white-spaces in the
non-JIT case which cause rebuilds.

Ma27 commented:
@K900 thanks again, excellent find, fixed that (and updated the tests to make sure we don't regress to the previous case).
@SuperSandro2000 I applied one of your suggestions as it might save us a few rebuilds (I guess we can re-target against `master` then, right @RaitoBezarius ?), thanks! Left a comment regarding the other.

also, pinging @marsam while this doesn't change the default behavior and Darwin is fixed now (and the runtime-closure's footprint is lower), it would be cool to get some feedback from you as you had some remarks re the previous attempts :)

RaitoBezarius commented:
> @K900 thanks again, excellent find, fixed that (and updated the tests to make sure we don't regress to the previous case). @SuperSandro2000 I applied one of your suggestions as it might save us a few rebuilds (I guess we can re-target against `master` then, right @RaitoBezarius ?), thanks! Left a comment regarding the other.

By running `nixpkgs-review` on this PR on x86_64-linux (with probably more cache than other people):

```
234 packages added: postgresql_jit.*
79 packages updated:
age age cstore_fdw-unstable cstore_fdw-unstable glom jsonb_deep_sum-unstable jsonb_deep_sum-unstable nixos-install-tools nominatim periods periods pg-safeupdate pg-safeupdate pgaudit pgaudit pgroonga pgroonga pgrouting pgrouting pgtap pgtap pgvector pgvector pg_auto_failover pg_auto_failover pg_bigm pg_bigm pg_cron pg_cron pg_ed25519 pg_ed25519 pg_hint_plan pg_hint_plan pg_hll pg_hll pg_ivm pg_ivm pg_partman pg_partman pg_rational pg_rational pg_relusage pg_relusage pg_repack pg_repack pg_similarity pg_similarity pg_topn pg_topn pipelinedb pipelinedb plpgsql_check plpgsql_check plr plr plv8 plv8 postgis postgis postgresql postgresql repmgr repmgr rum rum smlar-unstable smlar-unstable tds_fdw-unstable tds_fdw-unstable temporal_tables-unstable temporal_tables-unstable timescaledb timescaledb timescaledb_toolkit timescaledb_toolkit tsearch-extras tsearch-extras wal2json wal2json
```

Therefore, I do agree that we can retarget master again given the reasonable amount of rebuilds.

K900 commented:
Rebuilt from the latest version, JIT works, no mass rebuilds needed. Haven't really looked at the details of the change, but I'm not really qualified to comment on that anyway.

marsam reviewed (approved):
Sorry for the delay
LGTM overall. Thanks a lot for working on this!

I did test it on NixOS and Darwin, and it works fine 👍

Ma27 commented:
I think this is in a pretty usable state now, so I'll ask the other way round now, any *objections* to merge it as-is?

cc @SuperSandro2000 @mweinelt @RaitoBezarius @fpletz

Ma27 commented on pkgs/servers/sql/postgresql/default.nix:
```diff
@@ -204,50 +288,60 @@ let
+in self:
+  let packages = mkPackages self; in
+  packages
+  // self.lib.mapAttrs'
```
`genAttrs` is not applicable here trivially because you cannot modify the names with it, i.e. `genAttrs [ "foo" "bar" ] (const 1)` gives you `{ foo = 1; bar = 1; }`.

Also, I think that `genAttrs (map (x: "${x}_jit") (attrNames packages)) (postgres: ...)` is more complex than the current solution.

ajs124 reviewed (approved):
didn't test, but package changes look similar enough to what we're running and the rest looks reasonable as well

SuperSandro2000 commented:
Since this is not triggering rebuilds in libpq, we had a fair amount of reviews and hit is off by default, I am going ahead and merge it.
